File: 933.Number of Recent Calls/933.Number of Recent Calls.cpp
class RecentCounter {
public:
    vector<int> times;
    
    RecentCounter() {
        
    }
    
    int find_start(int x){
        int l = 0,r = times.size()-1;
        while(l < r){
            int mid = (l + r) / 2;
            if(times[mid] >= x) r = mid;
            else l = mid + 1;
        }
        return l;
    }
    
    int ping(int t) {
        times.push_back(t);
        if(t <= 3000) return times.size();
        int idx = find_start(t-3000);
        return times.size()-idx;
    }
};

File: 933.Number of Recent Calls/933.Number of Recent Calls.py
#------------deque实现。-------------------
from collections import deque
class RecentCounter:
    def __init__(self):
        self.pings = deque()        

    def ping(self, t):
        """
        :type t: int
        :rtype: int
        """
        self.pings.append(t)
        while self.pings[0] < t - 3000:
            self.pings.popleft()
        return len(self.pings)

		
#-------------二分搜索发优化搜索过程------------------------
class RecentCounter:
    def __init__(self):
        self.pings = list()       

    def ping(self, t):
        """
        :type t: int
        :rtype: int
        """
        self.pings.append(t)
        l, r = 0, len(self.pings)
        if t > 3000:
            while l < r:
                mid = (l + r)//2
                if self.pings[mid] < t - 3000:
                    l = mid + 1
                else:
                    r = mid
            
        return len(self.pings) - l
